### PR TITLE
Use CFG's internal region on CFGNodes

### PR DESCRIPTION
**Pull request by efferifick**
Base class CFGNode has two constructors,
1) one that takes a TR_Memory * parameter
2) one that takes a TR::Region& parameter
When the first constructor is used the sucessors and predecessors
are stored in the parameter region. Otherwise, they are stored in the
heap memory region.
Child class OMR::Block only implements the first type of constructor.
This commit creates another constructor for OMR::Block which takes a
TR::CFG & parameter and uses the internal memory region for calling
the second type of CFGNode constructor.

Signed-off-by: Erick Ochoa <eochoa@ualberta.ca>

Must be merged this OpenJ9 PR before https://github.com/eclipse/openj9/pull/6866

### CODE REVIEW

**Comment by efferifick** [comment]
@andrewcraik, TR::CFG is friends with OMR::Block as you suggested.

**Comment by andrewcraik** [comment]
I think this looks good - would appreciate a second opinion from @Leonardo2718

File: compiler/il/OMRBlock.hpp [hunk 1]
```diff
@@ -124,7 +127,8 @@ class OMR_EXTENSIBLE Block : public TR::CFGNode
       }
 
    virtual const char * getName(TR_Debug *);
-
+   void init(TR::TreeTop *entry, TR::TreeTop *exit);
```
Comment (Leonardo2718) [inline compiler/il/OMRBlock.hpp]: This init function should really be private.

File: compiler/infra/OMRCfg.hpp [hunk 2]
```diff
@@ -52,6 +52,7 @@ class TR_BlockCloner;
 class TR_BlockFrequencyInfo;
 class TR_ExternalProfiler;
 namespace TR { class Block; }
+namespace OMR { class Block; }
```
Comment (Leonardo2718) [inline compiler/infra/OMRCfg.hpp]: Is this forward declaration really needed? It seems... out of place.

Comment (efferifick) [reply compiler/infra/OMRCfg.hpp]: Hi, the forward declaration is indeed needed because otherwise the `friend` declaration will throw an error. I noticed that the only place where `OMR::Block` is forward declared is in [`./compiler/il/OMRBlock.hpp:30`](https://github.com/eclipse/omr/blob/979f5de3c93dbf8b98ab9cd24442773883ee9437/compiler/il/OMRBlock.hpp#L28) where it is surrounded by some ifdefs. Should I place this forward declaration surrounded by [these ifdefs](https://github.com/eclipse/omr/blob/979f5de3c93dbf8b98ab9cd24442773883ee9437/compiler/infra/OMRCfg.hpp#L30) macros as well?

Comment (Leonardo2718) [reply compiler/infra/OMRCfg.hpp]: Oh, that's right... because it's in a different namespace.

In that case, no. The extra ifdefs are not needed. However, I would prefer if this forward declaration was after all the `namespace TR` ones because it's kind of "special." A comment explaining why a forward declaration of something in the `OMR` namespace is needed would also be appropriate.

File: compiler/il/OMRBlock.hpp [hunk 3]
```diff
@@ -483,6 +486,7 @@ class OMR_EXTENSIBLE Block : public TR::CFGNode
     */
 
    private:
+   void init(TR::TreeTop *entry, TR::TreeTop *exit);
 
    void uncommonNodesBetweenBlocks(TR::Compilation *, TR::Block *, TR::ResolvedMethodSymbol *methodSymbol = NULL);
```
Comment (Leonardo2718) [inline compiler/il/OMRBlock.hpp]: `uncommonNodesBetweenBlocks()` is now private, but it used to be public. Was this change intentional? I would have expected this to break something.

File: compiler/infra/OMRCfg.hpp [hunk 4]
```diff
@@ -60,6 +60,10 @@ namespace TR { class ResolvedMethodSymbol; }
 namespace TR { class TreeTop; }
 template <class T> class TR_Array;
 
+// Needed for friend declaration
+// OMR::Block needs access to TR::Region _internalRegion in OMR::CFG
+namespace OMR { class Block; }
```
Comment (Leonardo2718) [inline compiler/infra/OMRCfg.hpp]: Out of curiosity, is there a reason for not using a public getter for `_internalRegion` instead of making `OMR::Block` a friend (assuming that's the only reason it's a friend)?

File: compiler/il/OMRBlock.hpp (hunk 3 above)
Comment (efferifick) [reply compiler/il/OMRBlock.hpp]: I didn't make `uncommonNodesBetweenBlocks` private. It is private on [master](https://github.com/eclipse/omr/blob/master/compiler/il/OMRBlock.hpp#L485) already.

File: compiler/infra/OMRCfg.hpp (hunk 4 above)
Comment (efferifick) [reply compiler/infra/OMRCfg.hpp]: I think it would be best not to make `_internalRegion` accessible through a public getter because that would allow the people to store things not related to CFG in `_internalRegion`.
But if you think it is a better idea than making it a friend, sure.

Comment (andrewcraik) [reply compiler/infra/OMRCfg.hpp]: I think either design would work so we can go with an API to get the region if that is cleaner, but make sure the API is get only with no setter so the allocation region stays consistent.

Comment (efferifick) [reply compiler/infra/OMRCfg.hpp]: So, just to be clear, I should change the API from a friend to a getter, right?

File: compiler/il/OMRBlock.hpp (hunk 3 above)
Comment (Leonardo2718) [reply compiler/il/OMRBlock.hpp]: Oh, you're right. The diff for this is a bit confusing.

File: compiler/infra/OMRCfg.hpp (hunk 4 above)
Comment (andrewcraik) [reply compiler/infra/OMRCfg.hpp]: Yes that is the conclusion of the review from @Leonardo2718 and I am fine with that

**Comment by efferifick** [comment]
@Leonardo2718 @andrewcraik Sorry for taking a bit to submit these changes! Let me know if there's anything else I should do.

**Comment by andrewcraik** [comment]
@genie-omr build all
